### ISSUE TITLE
Views exposed filters reset button doesn't work in a block

### ISSUE THREAD

**Issue by laryn** [backdrop/backdrop-issues#5531]
## Description of the bug

This D8 issue sounds remarkably similar to what I'm finding:

> ### Exposed filter reset redirects user to 404 page on AJAX view when placed as a block
> 
> https://www.drupal.org/project/drupal/issues/2820347
>
> Steps to recreate:
> 
> - Create an AJAX enabled view with exposed filters (as drop down), enable the reset button
> - Create a page and place the view as a block on the page.
> - Load the view page - no reset button
> - Click filter submit button
> - Reset button appears
> - Click Reset button
> 
> - The page is redirected to a system/404

In my case I'm being redirected back to the homepage when I press "Reset". I do have the "Inherits path" set to "Yes" and there are no page displays on this view, just block displays.

**Comment by JPustkuchen** [comment]
If someone runs into this, also see
https://www.drupal.org/project/drupal/issues/3346820
https://www.drupal.org/project/drupal/issues/3163940
https://www.drupal.org/project/drupal/issues/3335209
https://www.drupal.org/project/drupal/issues/2844823

**Comment by argiepiano** [comment]
@laryn, @JPustkuchen, is the issue happening only with AJAX enabled views? Or always?

**Comment by laryn** [comment]
I've only noticed it in relation to Ajax views -- in fact just ran into it again today. I'm seeing it with:

 - Views block with exposed filter, placed in a Layout
 - "Inherits path" is set to TRUE (also happens when set to FALSE, though)
 - Ajax is enabled
 - "Show reset button" is activated

When using the exposed filters, you can make changes and see results. When you click "Reset" however, the front page reloads.

Simply changing it to not use Ajax works -- albeit with full page reloads each time.

It looks like this issue (or one that is similar) is also present in modern Drupal: 

 - https://www.drupal.org/project/drupal/issues/3346820

And in Drupal 7:

 - https://julian.pustkuchen.com/en/drupal-7-views-3-add-ajaxified-reset-button-views-exposed-filters

**Aside:** Hey @JPustkuchen Is that link you? ☝️  😆 

## Possible solution?

I am wondering if we may have an opportunity to chop out some old code with long-standing issues. This is just a proof-of-concept, but in this section of the `views_plugin_exposed_form` class:

```php
  function exposed_form_alter(&$form, &$form_state) {
    if (!empty($this->options['reset_button'])) {
      $form['reset'] = array(
        '#value' => $this->options['reset_button_label'],
        '#type' => 'submit',
      );
    }
```

Instead of building the reset button as a `submit` I tried this manually using [type="reset"](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/input/reset):

```php
  function exposed_form_alter(&$form, &$form_state) {
    if (!empty($this->options['reset_button'])) {
      $form['reset'] = array('#markup' => '<input class="form-submit" type="reset" value="' . check_plain($this->options['reset_button_label']) . '" class="form-submit" />');
    }
```

And it seems to work. If we can do this in one way or another, there may be additional code we can chop out of `exposed_form_submit()` and `reset_form()` within this same class.

**Comment by indigoxela** [comment]
@laryn I'm having trouble to reproduce, so I guess, something's missing in the STR.

What I tried:

- A simple view from content (page)
- An exposed filter on content type
- Exposed filter in block
- Use ajax: yes
- Add a reset button (exposed form style settings)

Then I visited that page, edited "this page" layout to add the exposed block, back to page...

- Reset button was there
- Applied a filter (content type dropdown)
- Clicked the reset button and items got reset properly

What am I missing?

The fact that you land on the home page means, the reset button lost its JS event, I guess, but in my setup this doesn't happen.

**Comment by laryn** [comment]
@indigoxela Try the same, but in a view that has no page display at all, only blocks.  Then that block view with an exposed filter is placed on a page that is not the home page.

**Comment by indigoxela** [comment]
> Try the same, but in a view that has no page display at all, only blocks.

That's confusing... A block-only-view cannot have an exposed form in a block. There's not even a setting for that for blocks. Only _page views_ can have an exposed form in blocks. Are you trying to use the reset button from a page view on a block view?

**Comment by laryn** [comment]
I agree, it's confusing. It doesn't have "exposed filter in a block" the entire view is a block and also has an exposed filter. Here's a full config export in a fresh demo site, showing what I mean. You'll see it in the sidebar on all non-homepage pages.

[views-block-with-exposed-filter.tar.gz](https://github.com/user-attachments/files/21367083/views-block-with-exposed-filter.tar.gz)

**Comment by indigoxela** [comment]
Maybe, to make this all less confusing, provide a better issue description? The current one is *far* from ideal. 😆

**Comment by indigoxela** [comment]
Ahaaaaa... now I get it! Yes, I can reproduce that.

And as I suspected: the reset button doesn't have the "ajax" event attached (no event from ajax.js), hence resetting redirects to the home page.

**Comment by argiepiano** [comment]
I can also reproduce the problem. The reset button does a normal page request as opposed to an ajax one (which leads to the reload of the home page in some cases). This is true also for Views pages, not just blocks. The difference is that the Reset button in pages reloads the actual view, so the problem is less visible. 

As @indigoxela mentioned, Backdrop only attaches an ajax behavior to the first button. This is done in [this line](https://github.com/backdrop/backdrop/blob/003ccb38a84d7b1c6181d60d439524c9cf4d71be/core/modules/views/js/ajax_view.js#L97), where the code purposely only gets the first button of the form. 

We could try to iterate through all the buttons in the form, and attach the ajax behavior. Not sure if this may cause any issues. AFAIK an exposed form should not have more than two buttons (submit and reset), so iterating through its buttons should be enough. 

EDIT: I tried that, and was able to attach the ajax behavior successfully to both buttons, but the reset button actually didn't do its thing when operating through ajax - there is no change in the view, no errors. So, the fix is not as simple as just attaching the behavior. I will need to come back to this, or perhaps someone else can figure out a way to fi it before I do that.

**Comment by indigoxela** [comment]
@argiepiano many thanks for giving it a try.

Yes, the problem's a bit wider than initially assumed: exposed forms don't work properly on views blocks. Even the submit button fails without ajax. The reset button's obviously only considering page views, which have a path. And resetting loads the page without the filtering GET items (which won't work with ajax enabled).

The question's now: what to fix here. Block views with exposed forms without ajax (fix form action attribute value)? Ajaxify the reset button (if configured) and get that working? Switch to a different button type (type=reset)? Provide "type=reset" in core's Form API?

The problem has been reported for D7 views many, many times... there are several in this list: https://www.drupal.org/project/issues/views?text=exposed+form+block&status=Open&priorities=All&categories=All&version=All&component=All

**Comment by indigoxela** [comment]
Here's some starting point for further discussion: https://github.com/backdrop/backdrop/pull/5090

Note: this is quick and dirty and eventually the wrong spot (another option would be function views_exposed_form().

This fixes the form `#action` value, so it works properly for block views *without ajax* - no more redirect to home for submit and reset (also with pagers). I didn't dig too deep, though, so this might eventually have side effects somewhere.

Edit: a bit unexpected - all tests passing... 🙊

**Comment by indigoxela** [comment]
@laryn mind to update the issue description and title properly? According to our issue template.
Some copy-paste from a D10 issue isn't that helpful, even more as it's a thread with 147 comments. 😜

**Comment by argiepiano** [comment]
Hey @indigoxela. Thanks for the proof of concept. I see that this would skip ajax completely and reload the page, with the correct url where the block lives which is fine with me!  This looks like a simple solution. I'm a bit concerned about returning `current_path()` indiscriminately, though.

The Views object has the property `override_url`, which can be used to redirect to the correct url. This is used by some modules, like VBO (see this [block of code here](https://github.com/backdrop-contrib/views_bulk_operations/blob/f363e6e17b6c51fa422371d56e20855a9f81868a/views_bulk_operations.module#L407)), where the code first checks if `override_url` is set, in which case it respects it. If not, it essentially does what you are trying here, but it also attaches any path queries that exist. This apparently was done because of the issue we encountered in here (the page being redirected to the wrong url after submitting).

Do you think we could use a similar approach to avoid this problem? (i.e. checking if override_url is set, and if not, do the thing).

**Comment by argiepiano** [comment]
In a nutshell:

```php
  function get_url() {
    if ($this->get_option('inherit_path')) {
      return $this->get_path();
    }
    if (empty($this->view->override_url)) {
      $this->view->override_url = current_path();
    }
    return parent::get_url();
  }
```

I'm not sure if attaching an existing query is needed - more thinking and testing needed.

**Comment by indigoxela** [comment]
> I see that this would skip ajax completely and reload the page,

Yes, that's how the "Reset" button works by design. It strips off all GET and reloads the current page by calling the form action (url). And that url is "home" for blocks - they don't have paths (they're null), so url() ends up with home.

> I'm not sure if attaching an existing query is needed - more thinking and testing needed.

I belief, attaching the query breaks, how this button's supposed to work (GET should be stripped), but I need to take a closer look.

Using `parent::get_url();` will end up with home - that's why we need to get rid of it.
I'll try to clarify the problem in the issue description.

**Comment by indigoxela** [comment]
@argiepiano I'm playing locally with your suggestion... `$this->view->override_url` will always be null for block views - they're not supposed to have a path/url set.

So your suggestion would work, as we'd always get to `current_path()` by if clause.
However, I don't see, why we'd need the `parent::get_url();` fallback at all - it's always null for block views, which is never useful.

Hm... looking at VBO - they'll end up with the same, as `$_GET['q']` **is** the current path and nothing else. And the additional GET params are stripped off by intention - that's the "reset".
We don't have to do it in a "hackish" way as we're inside the class views_plugin_display_block and not in a hook like VBO.
Or am I missing something?

Edit: to clarify, why I belief, that the change has to happen in views_plugin_display_block::get_url(): That's used to set the form `#action` property - and that's where we end up when:

1. We submit an exposed form on a block view without ajax enabled
2. We submit the Reset button

It's the change with least impact on anything else. The next approach that comes to my mind is to completely change, how the Reset button works in all Views - and that might be too big of a change, might eventually require a major release. 😬

**Comment by argiepiano** [comment]
Let me explain my thinking, which may be wrong. The reason why I prefer to set  the  View's `override_url` and then call the parent's `get_url` is that:

1. Calling `parent::get_url()` does the minimum amount change to the functioning of original code, which called this method. My thinking is that this call to the parent method was there for a reason. Yes, I see the point that  `view::get_url()` will just returns the value of `override_url` when provided, so it would be reasonable to just return `current_path()` in the block's class instead of bothering using `override_url` and calling the parent's method. But who knows, in the future `view::get_url()` may change and do something different, some additional processing etc.  Completely removing the call to the parent's method in the block class would skip any other possibly important process that this method might do  
2. In my view, the property `override_url` exists for this type of cases when you need to specify an url different from the one provided by the view, OR to specify a url when none was there.
3. The reason for the check for `empty($this->view->override_url)` is that (maybe) there are some previous hooks that set override_url, or perhaps a module chooses to extend `views_plugin_display` for a new type of display based on the block class, and chooses to provide its own `override_url` value. If we don't check for empty, we run the risk of overwriting the correct url

I don't see my suggestion as "hackish". In fact, to me, calling the parent's homonymous methods is a very common OOP approach in well-designed subclasses. To me, removing this call is "hackish". 

Am I chasing ghosts? You know these classes better than me.

**Comment by indigoxela** [comment]
> Calling parent::get_url() does the minimum amount change to the functioning of original code

It's "null"... 🤷 And hence doesn't provide much value. But, sure, we can leave it there.

> The reason for the check for empty($this->view->override_url) is that (maybe) there are some previous hooks that set override_url, or perhaps a module chooses to extend...

That's a good point. [Looking at contrib](https://github.com/search?q=org%3Abackdrop-contrib%20override_url&type=code) - they usually override with either `current_path()` or with `$_GET['q']` directly.

It's a bit unfortunate, that the property isn't documented well in core - as so many times in Views.

> Am I chasing ghosts? You know these classes better than me.

It's always good to be cautious and to consider possible existing implementations for best BC. 😉 I don't think, I know these classes any better than you. 😀

**Comment by argiepiano** [comment]
> > Calling parent::get_url() does the minimum amount change to the functioning of original code

> It's  "null"... 🤷 And hence doesn't provide much value. But, sure, we can leave it there.

Can you clarify what is null?

**Comment by indigoxela** [comment]
> Can you clarify what is null?

parent::get_url() is null. There's a lot of inheritance in views - class views_plugin_display_block extends views_plugin_display, class views_plugin_display extends views_plugin, views_plugin extends views_object, but the path is actually fetched from the view object (class view)...

Path is null, unless modules override it, or unless one sets "Inherits path: Yes". 🤔 

Maybe that part works as designed. Views isn't always easy to understand.

IF "Inherits path" is meant to do exactly that - set a path from the page the block is attached, then I'd have to update the issue description (again). 😉

The reset works then for non-ajax blocks, but breaks with ajaxified blocks.

@argiepiano maybe it's better, if you take over here?

**Comment by argiepiano** [comment]
> parent::get_url() is null. 

Do you mean that the method `parent::get_url()` is null? Or that the path is null? Yes, the path is null, but the method is not null - it returns null at the moment, but that gets fix with your PR or with my suggested fix.

**Comment by indigoxela** [comment]
> Do you mean that the method parent::get_url() is null? 

No, of course not. What I actually meant: parent::get_url() **returns** null. 😉 Wording matters.

Edit: Maybe (busy) Mondays aren't the best day to discuss problems with Views. 🙊

**Comment by argiepiano** [comment]
> Edit: Maybe (busy) Mondays aren't the best day to discuss problems with Views. 🙊

🤷🏽 

I did not mean to derail your proposed fix or imply that I would take this over. Your fix Works For Me. My suggestion was just that, and I tried to justify it, but it seems you don't agree, which is fine - this is what Open Source is about.

Maybe @laryn can chime in.

I will not take this issue over.

**Comment by indigoxela** [comment]
> I did not mean to derail your proposed fix

The problem is: I'm not sure anymore. Not because of your input - I'm fine with it. It's because of some odd stuff I saw in views - which makes me consider my approach.

I'm not concerned, that you might take over - and I don't actually oppose to that `parent::get_url()`. I just don't get, why it's currently there (existing code).

**Comment by laryn** [comment]
My understanding of "Inherits path" is that it's for Views without a page display, and you want the block(s) to inherit the path of the page/layout that they are placed on rather than looking for a Views page display to attach to.

I did a quick test of the PR and when having "Inherits path" set and Ajax enabled, it still didn't work (reloaded the homepage instead). I'll try to test more thoroughly and provide an actual review this week.

**Comment by indigoxela** [comment]
> I did a quick test of the PR and when having "Inherits path" set and Ajax enabled, it still didn't work

Yes, that breaks because form action (path) gets wiped when it's AJAX, but reset buttons depend on that path.

> I'll try to test more thoroughly and provide an actual review this week.

Testing's welcome, and also feedback and hints re implementation, but it's actually too early for review. The current state "needs work" is set by intention - it needs work, it's just unclear in which direction this should go. 😉
